File: src/pdf/libqpdfselection.zig
const C = @import("qt6c");
const qclipboard_enums = @import("../libqclipboard.zig").enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qpdfselection.html
pub const qpdfselection = struct {
    /// New constructs a new QPdfSelection object.
    ///
    /// ``` other: ?*C.QPdfSelection ```
    pub fn New(other: ?*anyopaque) ?*C.QPdfSelection {
        return C.QPdfSelection_new(@ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpdfselection.html#operator=)
    ///
    /// ``` self: ?*C.QPdfSelection, other: ?*C.QPdfSelection ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        C.QPdfSelection_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpdfselection.html#swap)
    ///
    /// ``` self: ?*C.QPdfSelection, other: ?*C.QPdfSelection ```
    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        C.QPdfSelection_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpdfselection.html#isValid)
    ///
    /// ``` self: ?*C.QPdfSelection ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return C.QPdfSelection_IsValid(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpdfselection.html#text)
    ///
    /// ``` self: ?*C.QPdfSelection, allocator: std.mem.Allocator ```
    pub fn Text(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = C.QPdfSelection_Text(@ptrCast(self));
        defer C.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpdfselection.html#boundingRectangle)
    ///
    /// ``` self: ?*C.QPdfSelection ```
    pub fn BoundingRectangle(self: ?*anyopaque) ?*C.QRectF {
        return C.QPdfSelection_BoundingRectangle(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpdfselection.html#startIndex)
    ///
    /// ``` self: ?*C.QPdfSelection ```
    pub fn StartIndex(self: ?*anyopaque) i32 {
        return C.QPdfSelection_StartIndex(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpdfselection.html#endIndex)
    ///
    /// ``` self: ?*C.QPdfSelection ```
    pub fn EndIndex(self: ?*anyopaque) i32 {
        return C.QPdfSelection_EndIndex(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpdfselection.html#copyToClipboard)
    ///
    /// ``` self: ?*C.QPdfSelection ```
    pub fn CopyToClipboard(self: ?*anyopaque) void {
        C.QPdfSelection_CopyToClipboard(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpdfselection.html#copyToClipboard)
    ///
    /// ``` self: ?*C.QPdfSelection, mode: qclipboard_enums.Mode ```
    pub fn CopyToClipboard1(self: ?*anyopaque, mode: i64) void {
        C.QPdfSelection_CopyToClipboard1(@ptrCast(self), @intCast(mode));
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: ?*C.QPdfSelection ```
    pub fn QDelete(self: ?*anyopaque) void {
        C.QPdfSelection_Delete(@ptrCast(self));
    }
};

File: src/pdf/libqpdfselection.cpp
#include <QPdfSelection>
#include <QRectF>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qpdfselection.h>
#include "libqpdfselection.h"
#include "libqpdfselection.hxx"

QPdfSelection* QPdfSelection_new(const QPdfSelection* other) {
    return new QPdfSelection(*other);
}

void QPdfSelection_OperatorAssign(QPdfSelection* self, const QPdfSelection* other) {
    self->operator=(*other);
}

void QPdfSelection_Swap(QPdfSelection* self, QPdfSelection* other) {
    self->swap(*other);
}

bool QPdfSelection_IsValid(const QPdfSelection* self) {
    return self->isValid();
}

libqt_string QPdfSelection_Text(const QPdfSelection* self) {
    QString _ret = self->text();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

QRectF* QPdfSelection_BoundingRectangle(const QPdfSelection* self) {
    return new QRectF(self->boundingRectangle());
}

int QPdfSelection_StartIndex(const QPdfSelection* self) {
    return self->startIndex();
}

int QPdfSelection_EndIndex(const QPdfSelection* self) {
    return self->endIndex();
}

void QPdfSelection_CopyToClipboard(const QPdfSelection* self) {
    self->copyToClipboard();
}

void QPdfSelection_CopyToClipboard1(const QPdfSelection* self, int mode) {
    self->copyToClipboard(static_cast<QClipboard::Mode>(mode));
}

void QPdfSelection_Delete(QPdfSelection* self) {
    delete self;
}
